3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) 2014-2020 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.common.utils
+
+import cats.Monad
+import cats.data.{EitherT, NonEmptyList, Validated, ValidatedNel}
+import cats.effect.Clock
+import cats.implicits._
+
+import io.circe.Json
+
+import com.snowplowanalytics.iglu.client.{Client, ClientError}
+import com.snowplowanalytics.iglu.client.resolver.registries.RegistryLookup
+
+import com.snowplowanalytics.iglu.core.circe.instances._
+import com.snowplowanalytics.iglu.core.{SchemaCriterion, SchemaKey, SelfDescribingData}
+
+import com.snowplowanalytics.snowplow.badrows.FailureDetails
+
+import com.snowplowanalytics.snowplow.enrich.common.outputs.EnrichedEvent
+
+/** Contain the functions to validate:
+ *  - An unstructured event,
+ *  - The input contexts of an event,
+ *  - The contexts added by the enrichments.
+ */
+object IgluUtils {
+
+  /** Extract unstructured event from event and validate against its schema
+   *  @param event Snowplow event from which to extract unstructured event (in String)
+   *  @param client Iglu client used for SDJ validation
+   *  @param field Name of the field containing the unstructured event, to put in the bad row
+   *               in case of failure
+   *  @param criterion Expected schema for the JSON containing the unstructured event
+   *  @return Valid unstructured event if the input event has one
+   */
+  def extractAndValidateUnstructEvent[F[_]: Monad: RegistryLookup: Clock](
```

Comment:
No, I'd go with @benjben's approach. Its really easy to confuse them in many places. `Option` and `List` give us just enough protection.

Reply:
I think that having `extractAndValidateUnstructEvent` returning a `List` would be misleading as it would let us think that more than 1 element could be returned, whereas it will always be maximum one.